src/NITTA/LuaFrontend.hs

```diff
@@ -1,474 +1,377 @@
-{-# LANGUAGE AllowAmbiguousTypes #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE NamedFieldPuns #-}
 {-# LANGUAGE OverloadedStrings #-}
-{-# LANGUAGE PartialTypeSignatures #-}
-{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -Wno-type-defaults #-}
 
-{-# OPTIONS -fno-warn-type-defaults #-}
-
-{- |
-Module      : NITTA.LuaFrontend
-Description : Lua frontend prototype
-Copyright   : (c) Aleksandr Penskoi, 2019
-License     : BSD3
-Maintainer  : aleksandr.penskoi@gmail.com
-Stability   : experimental
--}
 module NITTA.LuaFrontend (
     lua2functions,
     FrontendResult (..),
     TraceVar (..),
+
+    -- * Internal
+    LuaAlgBuilder (..),
+    LuaStatement (..),
+    LuaValueVersion (..),
+    findStartupFunction,
+    getLuaBlockFromSources,
+    processStatement,
 ) where
 
-import Control.Monad.Identity
 import Control.Monad.State
-import Data.Bifunctor
 import qualified Data.HashMap.Strict as HM
-import Data.List (find, group, sort)
+import Data.Hashable
 import qualified Data.Map.Strict as M
 import Data.Maybe
 import Data.String
 import Data.String.ToString
-import qualified Data.String.Utils as S
-import Data.Text (Text, pack, unpack)
 import qualified Data.Text as T
 import Language.Lua
 import NITTA.Intermediate.DataFlow
 import qualified NITTA.Intermediate.Functions as F
-import NITTA.Utils (modify'_)
+import NITTA.Utils.Base
 import Text.Printf
 
+getUniqueLuaVariableName LuaValueVersion{luaValueVersionName, luaValueVersionIsConstant = True} luaValueAccessCount = "!" <> luaValueVersionName <> "#" <> showText luaValueAccessCount
+getUniqueLuaVariableName LuaValueVersion{luaValueVersionName, luaValueVersionAssignCount} luaValueAccessCount
+    | T.head luaValueVersionName == '_' = luaValueVersionName
+    | otherwise = luaValueVersionName <> "^" <> showText luaValueVersionAssignCount <> "#" <> showText luaValueAccessCount
+
+data LuaStatement x = LuaStatement
```

Comment:
In continue of https://github.com/ryukzak/nitta/pull/174#discussion_r681751624

The idea:
```
data LuaStatement x = LuaStatement
    { fArgs :: [LuaStatementArg x]
     , fOut :: [LuaValueVersion]
     , fName :: T.Text
     }
     deriving (Show, Eq)

data LuaStatementArg x 
  = LuaValueArg T.Text
  | InternediateArg x
  | IntArg Int
  | …

…
function2nitta LuaStatement{fName = "loop", fIn = [a], fArgs = [LuaValueArg c, InternediateArg c]} = F.loop x a $ output c
…      
```

I think such way is more maintainable.

Reply:
this will make the pattern matching extremely large.
for example, the addition operation is available for any combination of an integer, a fractional number and a Lua variable, which means that only for addition we will need to register 9 different patterns, one for every possible combination.
Do you think it's worth it?

Reply:
I think my idea base on a leak from the previous implementation. Early, a statement like `x+1` we rewrote to `x+const_1`. So it was an application for two variables, and the complexity problem does not exist early.

Also, we don't need to list all possible options as a pattern because we can check their correctness in a general way.

Also, maybe I find an issue if I correctly catch one detail. Are statements for `1 << x` and `x << 1` equal?

Reply:
Please, extract it to separated issue.

Reply:
https://github.com/ryukzak/nitta/issues/185